File: src/lib/wow/formats/adt/wdt-loader.ts
/**
 * WDT loader, ported from wow.export (src/js/3D/loaders/WDTLoader.js).
 */
import { BufferWrapper } from '../buffer';
import { constants } from '../constants';

const MAP_SIZE = constants.GAME.MAP_SIZE;
const MAP_SIZE_SQ = constants.GAME.MAP_SIZE_SQ;

export interface WDTEntry {
  rootADT: number;
  obj0ADT: number;
  obj1ADT: number;
  tex0ADT: number;
  lodADT: number;
  mapTexture: number;
  mapTextureN: number;
  minimapTexture: number;
}

export interface WorldModelPlacement {
  id: number;
  uid: number;
  position: number[];
  rotation: number[];
  upperExtents: number[];
  lowerExtents: number[];
  flags: number;
  doodadSetIndex: number;
  nameSet: number;
  padding: number;
}

export class WDTLoader {
  data: BufferWrapper;

  flags?: number;

  lgtFileDataID?: number;

  occFileDataID?: number;

  fogsFileDataID?: number;

  mpvFileDataID?: number;

  texFileDataID?: number;

  wdlFileDataID?: number;

  pd4FileDataID?: number;

  tiles?: number[];

  entries?: WDTEntry[];

  worldModel?: string;

  worldModelPlacement?: WorldModelPlacement;

  constructor(data: BufferWrapper) {
    this.data = data;
  }

  /** Load the WDT file, parsing it. */
  load(): void {
    while (this.data.remainingBytes > 0) {
      const chunkID = this.data.readUInt32LE();
      const chunkSize = this.data.readUInt32LE();
      const nextChunkPos = this.data.offset + chunkSize;

      const handler = WDTChunkHandlers[chunkID];
      if (handler) handler.call(this, this.data, chunkSize);

      // Ensure that we start at the next chunk exactly.
      this.data.seek(nextChunkPos);
    }
  }
}

const WDTChunkHandlers: Record<number, (this: WDTLoader, data: BufferWrapper, chunkSize: number) => void> = {
  // MPHD (Flags)
  0x4D504844(data) {
    this.flags = data.readUInt32LE();
    this.lgtFileDataID = data.readUInt32LE();
    this.occFileDataID = data.readUInt32LE();
    this.fogsFileDataID = data.readUInt32LE();
    this.mpvFileDataID = data.readUInt32LE();
    this.texFileDataID = data.readUInt32LE();
    this.wdlFileDataID = data.readUInt32LE();
    this.pd4FileDataID = data.readUInt32LE();
  },

  // MAIN (Tiles)
  0x4D41494E(data) {
    const tiles = new Array<number>(MAP_SIZE_SQ);
    this.tiles = tiles;
    for (let x = 0; x < MAP_SIZE; x++) {
      for (let y = 0; y < MAP_SIZE; y++) {
        tiles[(y * MAP_SIZE) + x] = data.readUInt32LE();
        data.move(4);
      }
    }
  },

  // MAID (File IDs)
  0x4D414944(data) {
    const entries = new Array<WDTEntry>(MAP_SIZE_SQ);
    this.entries = entries;

    for (let x = 0; x < MAP_SIZE; x++) {
      for (let y = 0; y < MAP_SIZE; y++) {
        entries[(y * MAP_SIZE) + x] = {
          rootADT: data.readUInt32LE(),
          obj0ADT: data.readUInt32LE(),
          obj1ADT: data.readUInt32LE(),
          tex0ADT: data.readUInt32LE(),
          lodADT: data.readUInt32LE(),
          mapTexture: data.readUInt32LE(),
          mapTextureN: data.readUInt32LE(),
          minimapTexture: data.readUInt32LE(),
        };
      }
    }
  },

  // MWMO (World WMO)
  0x4D574D4F(data, chunkSize) {
    this.worldModel = data.readString(chunkSize).replace('\0', '');
  },

  // MODF (World WMO Placement)
  0x4D4F4446(data) {
    this.worldModelPlacement = {
      id: data.readUInt32LE(),
      uid: data.readUInt32LE(),
      position: data.readFloatLE(3),
      rotation: data.readFloatLE(3),
      upperExtents: data.readFloatLE(3),
      lowerExtents: data.readFloatLE(3),
      flags: data.readUInt16LE(),
      doodadSetIndex: data.readUInt16LE(),
      nameSet: data.readUInt16LE(),
      padding: data.readUInt16LE(),
    };
  },
};

export default WDTLoader;

File: golang/internal/wow/formats/adt/wdt_loader.go
package adt

import (
	"github.com/pqhuy98/wow-converter/internal/buffer"
	"github.com/pqhuy98/wow-converter/internal/wow/constants"
)

// WDTEntry holds file data IDs for a map tile.
type WDTEntry struct {
	RootADT         uint32
	Obj0ADT         uint32
	Obj1ADT         uint32
	Tex0ADT         uint32
	LodADT          uint32
	MapTexture      uint32
	MapTextureN     uint32
	MinimapTexture  uint32
}

// WorldModelPlacement describes a world WMO placement from WDT.
type WorldModelPlacement struct {
	ID              uint32
	UID             uint32
	Position        [3]float32
	Rotation        [3]float32
	UpperExtents    [3]float32
	LowerExtents    [3]float32
	Flags           uint16
	DoodadSetIndex  uint16
	NameSet         uint16
	Padding         uint16
}

// WDTLoader parses .wdt map descriptor files.
type WDTLoader struct {
	Data *buffer.Buffer

	Flags                uint32
	LgtFileDataID        uint32
	OccFileDataID        uint32
	FogsFileDataID       uint32
	MpvFileDataID        uint32
	TexFileDataID        uint32
	WdlFileDataID        uint32
	Pd4FileDataID        uint32
	Tiles                []uint32
	Entries              []WDTEntry
	WorldModel           string
	WorldModelPlacement  *WorldModelPlacement
}

// NewWDTLoader creates a WDT loader.
func NewWDTLoader(data *buffer.Buffer) *WDTLoader {
	return &WDTLoader{Data: data}
}

// Load parses the WDT file.
func (w *WDTLoader) Load() {
	for w.Data.RemainingBytes() > 0 {
		chunkID := readU32(w.Data)
		chunkSize := int(readU32(w.Data))
		nextChunkPos := w.Data.Offset() + chunkSize
		if handler, ok := wdtChunkHandlers[chunkID]; ok {
			handler(w, w.Data, chunkSize)
		}
		w.Data.Seek(nextChunkPos)
	}
}

type wdtChunkHandler func(*WDTLoader, *buffer.Buffer, int)

var wdtChunkHandlers = map[uint32]wdtChunkHandler{
	0x4D504844: func(w *WDTLoader, data *buffer.Buffer, _ int) {
		w.Flags = readU32(data)
		w.LgtFileDataID = readU32(data)
		w.OccFileDataID = readU32(data)
		w.FogsFileDataID = readU32(data)
		w.MpvFileDataID = readU32(data)
		w.TexFileDataID = readU32(data)
		w.WdlFileDataID = readU32(data)
		w.Pd4FileDataID = readU32(data)
	},
	0x4D41494E: func(w *WDTLoader, data *buffer.Buffer, _ int) {
		tiles := make([]uint32, constants.Game.MapSizeSq)
		w.Tiles = tiles
		for x := 0; x < constants.Game.MapSize; x++ {
			for y := 0; y < constants.Game.MapSize; y++ {
				tiles[y*constants.Game.MapSize+x] = readU32(data)
				data.Move(4)
			}
		}
	},
	0x4D414944: func(w *WDTLoader, data *buffer.Buffer, _ int) {
		entries := make([]WDTEntry, constants.Game.MapSizeSq)
		w.Entries = entries
		for x := 0; x < constants.Game.MapSize; x++ {
			for y := 0; y < constants.Game.MapSize; y++ {
				entries[y*constants.Game.MapSize+x] = WDTEntry{
					RootADT: readU32(data), Obj0ADT: readU32(data), Obj1ADT: readU32(data),
					Tex0ADT: readU32(data), LodADT: readU32(data), MapTexture: readU32(data),
					MapTextureN: readU32(data), MinimapTexture: readU32(data),
				}
			}
		}
	},
	0x4D574D4F: func(w *WDTLoader, data *buffer.Buffer, chunkSize int) {
		w.WorldModel = stringsTrimNull(data.ReadString(chunkSize, "utf8"))
	},
	0x4D4F4446: func(w *WDTLoader, data *buffer.Buffer, _ int) {
		pos := readFloat3(data)
		rot := readFloat3(data)
		upper := readFloat3(data)
		lower := readFloat3(data)
		w.WorldModelPlacement = &WorldModelPlacement{
			ID: readU32(data), UID: readU32(data),
			Position: pos, Rotation: rot, UpperExtents: upper, LowerExtents: lower,
			Flags: uint16(readU16(data)), DoodadSetIndex: uint16(readU16(data)),
			NameSet: uint16(readU16(data)), Padding: uint16(readU16(data)),
		}
	},
}

func stringsTrimNull(s string) string {
	if i := stringsIndexByte(s, 0); i >= 0 {
		return s[:i]
	}
	return s
}

func stringsIndexByte(s string, b byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == b {
			return i
		}
	}
	return -1
}

func readU32(b *buffer.Buffer) uint32 { return uint32(b.ReadUInt32LE().(int64)) }

func readI64(b *buffer.Buffer) int64 { return b.ReadInt64LE().(int64) }

func readU16(b *buffer.Buffer) uint16 { return uint16(b.ReadUInt16LE().(int64)) }

func readFloat3(b *buffer.Buffer) [3]float32 {
	raw := b.ReadFloatLE(3).([]float32)
	return [3]float32{raw[0], raw[1], raw[2]}
}
